homeassistant/components/klyqa/__init__.py

```diff
@@ -0,0 +1,137 @@
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up or change Klyqa integration from a config entry."""
+    username = str(entry.data[CONF_USERNAME])
+    password = str(entry.data.get(CONF_PASSWORD))
+    host = str(entry.data.get(CONF_HOST))
+    scan_interval = int(entry.data[CONF_SCAN_INTERVAL])
+    polling = bool(entry.data.get(CONF_POLLING))
+
+    sync_rooms: bool = (
+        entry.data[CONF_SYNC_ROOMS] if entry.data[CONF_SYNC_ROOMS] else False
+    )
+    component: KlyqaDataCoordinator = hass.data[DOMAIN]
+    component.scan_interval = timedelta(seconds=scan_interval)
+
+    klyqa_api: HAKlyqaAccount
+    if (
+        DOMAIN in hass.data
+        and hasattr(component, "entries")
+        and entry.entry_id in component.entries
+    ):
+        klyqa_api = component.entries[entry.entry_id]
+        await hass.async_add_executor_job(klyqa_api.shutdown)
+
+        klyqa_api.username = username
+        klyqa_api.password = password
+        klyqa_api.host = host
+        klyqa_api.sync_rooms = sync_rooms
+        klyqa_api.polling = polling
+        klyqa_api.scan_interval = scan_interval
+    else:
+        klyqa_api = HAKlyqaAccount(
+            component.udp,
+            component.tcp,
+            username,
+            password,
+            host,
+            hass,
+            sync_rooms=sync_rooms,
+            polling=polling,
+            scan_interval=scan_interval,
+        )
+        if not hasattr(component, "entries"):
+            component.entries = {}
+        component.entries[entry.entry_id] = klyqa_api
+
+    if not await klyqa_api.login():
+        return False
+
+    entry.async_on_unload(entry.add_update_listener(async_reload_entry))
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, klyqa_api.shutdown)
+
+    # For previous config entries where unique_id is None
+    if entry.unique_id is None:
+        hass.config_entries.async_update_entry(
+            entry, unique_id=entry.data[CONF_USERNAME]
+        )
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+    if not unload_ok:
+        return unload_ok
+
+    while hass.data[DOMAIN].remove_listeners:
+        listener = hass.data[DOMAIN].remove_listeners.pop(-1)
+        try:
+            listener()
+        except:  # noqa: E722 pylint: disable=bare-except
+            pass
+
+    if DOMAIN in hass.data:
+        if entry.entry_id in hass.data[DOMAIN].entries:
+            if hass.data[DOMAIN].entries[entry.entry_id]:
+                await hass.async_add_executor_job(
+                    hass.data[DOMAIN].entries[entry.entry_id].shutdown
+                )
+            hass.data[DOMAIN].entries.pop(entry.entry_id)
```

Comment:
Not sure if I follow this one, this shuts down all config entries? But this method should only clean up a single entry.

Reply:
as it does right?
Can't spot a loop there.
It takes the entry_id from the entry and parses it to our according account in our api for a shutdown.
So far everything is fine from my point of view.